README.md

```diff
@@ -160,21 +178,111 @@ try {
+
+console.log(a, b, c);
+
+});
+```
+
+Modules `a.mjs`, `b.mjs`, and `c.mjs` would all execute in order up until the first await in each of them; we then wait on all of them to resume and finish evaluating before continuing.
+
 ### FAQ
 
 #### Isn't top-level `await` a footgun?
 
-If you have seen [the gist](https://gist.github.com/Rich-Harris/0b6f317657f5167663b493c722647221) you likely have heard this critique before. My hope is that as a committee we can weigh the pros / cons of the various approaches and determine if the benefits of the feature outweigh the risks.
+If you have seen [the gist](https://gist.github.com/Rich-Harris/0b6f317657f5167663b493c722647221) you likely have heard this critique before.
+
+Some responses to some of the top concerns here:
+
+##### Will top-level `await` cause developers to make their code block longer than it should?
+
+It's true that top-level `await` gives developers a new tool to make their code wait. Our hope is that proper developer education can ensure that the semantics of top-level `await` are well-understood, so that people know to use it just when they intend that importers should block on it.
+
+We've seen this work well in the past. For example, it's easy to write code with async/await that serializes two tasks that could be done in parallel, but a deliberate developer education effort has popularized the use of `Promise.all` to avoid this hazard.
+
+##### Will top-level `await` encourage developers to use `import()` unnecessarily, which is less optimizable?
+
+Many JavaScript developers are learning about `import()` specifically as a tool for code splitting. People are becoming aware of the relationship between bundling and multiple requests, and learning how to combine them for good application performance. Top-level `await` doesn't really change the calculus--using `import()` from a top-level `await` will have similar performance effects to using it from a function. As long as we can tie top-level `await`'s educational materials into the existing knowledge of that performance tradeoff, we hope to be able to avoid counterproductive increases in the use of `import()`.
+
+#### What exactly is blocked by a top-level `await`?
+
+When one module imports another one, the importing module will only start executing its module body once the dependency's body has finished executing. If the dependency reaches a top-level await, that will have to complete before the importing module's body starts executing.
+
+#### Why doesn't top-level `await` block the import of an adjacent module?
+
+If one module wants to declare itself dependent on another module, for the purposes of waiting for that other module to complete its top-level `await` statements before the module body executes, it can declare that other module as an import.
+
+In a case such as the following, the printed order will be `"X1"`, `"Y"`, `"X2"`, because importing one module "before" another does not create an implicit dependency.
```

Comment:
It should - no, **must** - stay the same, as there is no `await` anywhere in the example and we want to be backwards-compatible.
I think with `Promise.all` semantics, the modules will load in the same order as previously, will wait for their dependencies (which are synchronous, so immediately fulfilled promises), and then execute in that same order as promise tasks run in the same order they were queued in. The only difference is that they will be run in microtasks, see #43, but that doesn't affect order.
Oh, wait, you are concerned that `a` needs to wait for more promise resolution stuff than `b`, and might be evaluated after it?

Reply:
Right, the semantics of this proposal are that the order would stay the same as what you describe. The Promise.all affects how JS waits for dependencies, not how it launches their execution. (However, there is an observable change in that some turn(s) through the job queue would be taken in between modules.)

Seems like this property has been confusing and difficult to explain. Your help in pinpointing what causes this confusion would be appreciated.

Reply:
@bergus That's right. It seems from the explainer that the result will now be:

```
a-1
a-2
b
a
main
```

Because "a" has to defer it's body evaluation until Promise all'ing "a-1" and "a-2", which will push its evaluation job out past "b".

Does that make sense?

Reply:
@littledan not sure whether it's just a mistake in my naive attempt at translating @zenparsing's example to the "`Promise.all` semantics" and the actual semantics in #33 handle this correctly (I haven't dived into the details yet), or whether it is an actual issue, but
```js
const a1 = Promise.resolve()    .then(() => console.log('a-1'));
const a2 = Promise.resolve()    .then(() => console.log('a-2'));
const a  = Promise.all([a1, a2]).then(() => console.log('a'));
const b  = Promise.resolve()    .then(() => console.log('b'));
           Promise.all([a, b])  .then(() => console.log('main'));
```
does output the logs in the wrong order due to `Promise.all([a1, a2])` queuing more promise jobs than the `Promise.resolve` that `b` waits for.

Reply:
@zenparsing Oh, do you mean it would be pushed out because of several turns through the job queue would make "a" land later than "b" (in a sort of "race") where the delay isn't really triggered by waiting external resources? Maybe what we're stumbling towards is that we should drain the microtask queue between sibling execution!

Reply:
Maybe, but explicit queue draining would be a wart imo.
I guess it's easier to have a "synchronous" version of `Promise.all`, that checks whether all of its arguments are fulfilled already and if so immediately returns a result, falling back to the standard asynchronous procedure otherwise. (Yes, I know about Zalgo, but it could be warranted here). Even better if made explicit by using a result type of `SynchronouslyEvaluated<ModuleRecord> | TlaEvaluated<Promise<ModuleRecord>>`.

The alternative would be to make the traversal explicit and maintain a graph or queue, instead of doing it implicitly in the recursive algorithm. This might even give way to explicitly allow for concurrent (pre)loading of module code.

Reply:
Let's continue this discussion in https://github.com/tc39/proposal-top-level-await/issues/47 . Variant B has been proposed for a long time, and isn't introduced by this PR.